pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go

```diff
@@ -0,0 +1,142 @@
+package clientmiddleware
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/services/caching"
+	"github.com/grafana/grafana/pkg/services/contexthandler"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// NewCachingMiddleware creates a new plugins.ClientMiddleware that will
+// attempt to read and write query results to the cache
+func NewCachingMiddleware(cachingService caching.CachingService) plugins.ClientMiddleware {
+	log := log.New("caching_middleware")
+	if err := prometheus.Register(QueryRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector 'QueryRequestHistogram'", "error", err)
+	}
+	if err := prometheus.Register(ResourceRequestHistogram); err != nil {
+		log.Error("error registering prometheus collector 'ResourceRequestHistogram'", "error", err)
+	}
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &CachingMiddleware{
+			next:    next,
+			caching: cachingService,
+			log:     log,
+		}
+	})
+}
+
+type CachingMiddleware struct {
+	next    plugins.Client
+	caching caching.CachingService
+	log     log.Logger
+}
+
+// QueryData receives a data request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the queries as usual, then write the response to the cache.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if req == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	reqCtx := contexthandler.FromContext(ctx)
+	if reqCtx == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	// time how long this request takes
+	start := time.Now()
+
+	// First look in the query cache if enabled
+	hit, cr := m.caching.HandleQueryRequest(ctx, req)
+
+	defer func() {
+		// record request duration if caching was used
+		if ch := reqCtx.Resp.Header().Get(caching.XCacheHeader); ch != "" {
```

Comment:
Looks a bit weird looking at HTTP response header. Doesn't this depend on the value of `hit`? Is it because that response header only is set if enterprise and query caching enabled?

Reply:
@PoorlyDefinedBehaviour shared a similar opinion above, so I changed the interface to return a bool instead [1bb82aa](https://github.com/grafana/grafana/pull/65616/commits/1bb82aa25e5da0d997a11adbf0341e6454bbea68). OSS no longer cares what headers are set.